### DISCUSSION TITLE
AssetIn(key=[...]) Dependency building for multiple assets using

### DISCUSSION THREAD

**Discussion by raiAmagi** [dagster-io/dagster#18918] [Q&A] [answered]
I am reading and studying the documentation to use Dagster.

In the "Defining explicit managed-loading dependencies" section of the "Software-defined assets" page, there is the following code description
@asset(ins={"upstream": AssetIn(key=["some_db_schema", "upstream_asset"])})

So I created an asset.py with the following code and loaded it in DagsterUI, but it raises an exception.

Code:

from dagster import AssetIn, asset

@asset
def some_db_schema():
    return [1, 2, 3]

@asset
def upstream_asset():
    return [1, 2, 3]

@asset(ins={"upstream": AssetIn(key=["upstream_asset", "some_db_schema"])})
def another_downstream_asset(upstream):
    return upstream + [10]

Exception:
DagsterInvalidDefinitionError: Input asset '["upstream_asset", "some_db_schema"]' for asset '["another_downstream_ asset"]]' is not produced by any of the provided asset ops and is not one of the provided sources

I believe the problem is with the upstream asset.

Why is this an exception in my code?
What code should I put in AssetIn(key=[...]) What code should I put in the AssetIn(key=[...]) code to make the code work correctly?

The execution environment is as follows
os: windows
python: 3.11.6
dagster: 1.5.13

I have a problem with upstream assets.
I am assuming that DagsterUI will show the same dependencies as in the following two types of code.

Pattern 1:
@asset
def another_downstream_args_asset(upstream_asset, some_db_schema):.
　...
Pattern 2:
@asset(deps=[upstream_asset, some_db_schema])
def another_downstream_deps_asset():.
　...

**Comment by johannkm** [comment, marked as answer]
`AssetIn(key=["some_db_schema", "upstream_asset"])` actually specifies one asset key with the prefix `some_db_schema`. See the `key_prefix` argument to [@asset](https://docs.dagster.io/_apidocs/assets#dagster.asset).

To specify two dependencies, it'd be

```python
@asset
def some_db_schema():
  return [1, 2, 3]

@asset
def upstream_asset():
  return [1, 2, 3]

@asset(ins={"upstream_1": AssetIn(key=["upstream_asset"]), "upstream_2": AssetIn(key=["some_db_schema"])})
def another_downstream_asset(upstream_1, upstream_2):
  return upstream + [10]
```

**Reply by raiAmagi** [reply]
Thank you.

I am thinking that both should work the same as key_prefix and use the following.

・If you want to give the contents of an asset whose key is a dependent asset as an argument to a downstream function, use "ins={"upstream": AssetIn(key=...)". is used.

・If you do not need to pass the contents to the downstream function argument, use key_prefix.

Thank you for telling me about the dependency setting I had assumed.
I will try it as soon as possible.

**Reply by aviral-nayya** [reply]
Does this still work if the upstream assets are being imported from another module?

**Reply by garethbrickman** [reply]
@aviral-nayya You won't need to import the assets from another module, the asset key will resolve across the entire project (this is why asset keys should always be globally unique)

**Reply by fredonia88** [reply]
Is there a way to access the entire `ins` object in `another_downstream_asset`? Say `upstream_1` and `upstream_2` emit an s3_path. I want those s3 paths contained in one object. Something like: `{'upstream_1': upstream_1, 'upstream_2': upstream_2}`. Is this possible?

**Reply by garethbrickman** [reply]
@fredonia88 Does combining them in the downstream asset's code work for that? 

```python
from dagster import AssetIn, asset, AssetExecutionContext, Definitions

@asset
def upstream_asset_1():
    return "s3://bucket/path1"

@asset
def upstream_asset_2():
    return "s3://bucket/path2"

@asset(ins={"upstream_1": AssetIn(key=["upstream_asset_1"]), "upstream_2": AssetIn(key=["upstream_asset_2"])})
def another_downstream_asset(context, upstream_1, upstream_2):
    combined_paths = {
        "upstream_1": upstream_1,
        "upstream_2": upstream_2
    }
    context.log.info(f"Combined paths: {combined_paths}")
    return combined_paths

defs = Definitions(
    assets=[upstream_asset_1, upstream_asset_2, another_downstream_asset]
)
```
